feat: DIA-1422: Relationship between prompt version and provider conn…

hakan458 opened the pull request:
…ection

### PR fulfills these requirements
- [ ] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [ ] Self-reviewed and ran all changes on a local instance (for bug fixes/features)



#### Change has impacts in these area(s)
_(check all that apply)_
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [ ] Frontend



### Describe the reason for change
_(link to issue, supportive screenshots etc.)_



#### What does this fix?
_(if this is a bug fix)_



#### What is the new behavior?
_(if this is a breaking or feature change)_



#### What is the current behavior?
_(if this is a breaking or feature change)_



#### What libraries were added/updated?
_(list all with version changes)_



#### Does this change affect performance?
_(if so describe the impacts positive or negative)_



#### Does this change affect security?
_(if so describe the impacts positive or negative)_



#### What alternative approaches were there?
_(briefly list any if applicable)_



#### What feature flags were used to cover this change?
_(briefly list any if applicable)_



### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [ ] Not sure (briefly explain the situation below)



### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [ ] unit



### Which logical domain(s) does this change affect?
_(for bug fixes/features, be as precise as possible. ex. Authentication, Annotation History, Review Stream etc.)_


<!-- ELLIPSIS_HIDDEN -->


----

| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit 7280de5ef2ab53b995d4635b4e62f59049d77f15  | 
|--------|--------|

feat: DIA-1422: Add relationship between ThirdPartyModelVersion and ModelProviderConnection

### Summary:
Adds a foreign key to `ThirdPartyModelVersion` for `ModelProviderConnection`, updates `label-studio-sdk`, and includes a checklist for PR requirements.

**Key points**:
- Adds `model_provider_connection` foreign key to `ThirdPartyModelVersion` in `models.py`.
- Adds migration `0011_thirdpartymodelversion_model_provider_connection.py` to add and populate `model_provider_connection` field.
- Handles `OPENAI` and `AZURE_OPENAI` providers specifically in migration script.
- Updates `label-studio-sdk` dependency in `pyproject.toml`.
- Includes checklist for PR requirements and impact areas.


----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)


----


<!-- ELLIPSIS_HIDDEN -->
- https://github.com/HumanSignal/label-studio-sdk/pull/306 linked by [Workflow Run](https://github.com/HumanSignal/label-studio/actions/runs/10842719671)

Commit:
feat: DIA-1422: Relationship between prompt version and provider connection

Commit:
[submodules] Bump HumanSignal/label-studio-sdk version

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10842719671

Commit:
blue

Commit:
Merge branch 'fb-dia-1422' of github.com:heartexlabs/label-studio into fb-dia-1422

niklub commented on label_studio/ml_models/migrations/0011_thirdpartymodelversion_model_provider_connection.py:
```diff
@@ -0,0 +1,20 @@
+import django.db.models.deletion
+
+
+class Migration(migrations.Migration):
```
To make it backward compatible, we need to add migrations that would relate all previously existing model connections to the model versions. Then we can safely get rid of the current logic in the code

niklub replied:
Check this as an example https://github.com/HumanSignal/label-studio/blob/develop/label_studio/data_manager/migrations/0010_auto_20230718_1423.py#L42

Commit:
fill new column in migration (async)

fern-api[bot] committed:
[submodules] Bump HumanSignal/label-studio-sdk version

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10853581353

matt-bernstein commented on label_studio/ml_models/models.py:
```diff
@@ -67,6 +67,10 @@ class Meta:
 
     prompt = models.TextField(_('prompt'), null=False, blank=False, help_text='Prompt to execute')
 
+    model_provider_connection = models.ForeignKey(
+        ModelProviderConnection, related_name='model_versions', on_delete=models.SET_NULL, null=True
+    )
+
```
is there a way to link the `provider` attr in ThirdPartyModelVersion to this object? migration to make it a `@property`?

hakan458 replied:
Link provider to ModelVersion? Not sure what you mean or what you want to achieve here

niklub commented on label_studio/ml_models/migrations/0011_thirdpartymodelversion_model_provider_connection.py:
```diff
@@ -0,0 +1,54 @@
+    start_job_async_or_sync(_fill_model_version_model_provider_connection)
+
+
+def backwards(apps, schema_editor):
```
how about backward migrations? @farioas will be 😞😞😞

hakan458 replied:
Dont need them in this case since if we revert this migration, the column is just removed. Its in the same migration file

fern-api[bot] committed:
[submodules] Bump HumanSignal/label-studio-sdk version

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10854080761

hakan458 committed:
[submodules] Bump HumanSignal/label-studio-sdk version

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10854864553

Commit:
atomic=False for migration

Commit:
Merge branch 'fb-dia-1422' of github.com:heartexlabs/label-studio into fb-dia-1422

jombooth commented on label_studio/ml_models/migrations/0011_thirdpartymodelversion_model_provider_connection.py:
```diff
@@ -0,0 +1,57 @@
+    third_party_model_versions = ThirdPartyModelVersion.objects.all()
+    for model_version in third_party_model_versions:
+        if model_version.provider == ModelProviders.OPENAI:
+            with transaction.atomic():
```
Why do we need this?

jombooth commented on label_studio/ml_models/migrations/0011_thirdpartymodelversion_model_provider_connection.py:
```diff
@@ -0,0 +1,57 @@
+                    provider=model_version.provider
+                ).first()
+                model_version.model_provider_connection = model_provider_connection
+                model_version.save()
```
recommend `update_fields` here

jombooth commented on label_studio/ml_models/migrations/0011_thirdpartymodelversion_model_provider_connection.py:
```diff
@@ -0,0 +1,57 @@
+                    deployment_name=model_version.provider_model_id,
+                ).first()
+                model_version.model_provider_connection = model_provider_connection
+                model_version.save()
```
recommend `update_fields` here too

jombooth commented on label_studio/ml_models/migrations/0011_thirdpartymodelversion_model_provider_connection.py:
```diff
@@ -0,0 +1,57 @@
+                model_version.model_provider_connection = model_provider_connection
+                model_version.save()
+        elif model_version.provider == ModelProviders.AZURE_OPENAI:
+            with transaction.atomic():
```
I also don't understand why we need this transaction

jombooth commented on label_studio/ml_models/migrations/0011_thirdpartymodelversion_model_provider_connection.py:
```diff
@@ -0,0 +1,57 @@
+                model_provider_connection = ModelProviderConnection.objects.filter(
+                    organization=model_version.organization,
+                    provider=model_version.provider
+                ).first()
```
please use `fast_first` here and below

Commit:
ignore migration in linter

jombooth commented on label_studio/ml_models/migrations/0011_thirdpartymodelversion_model_provider_connection.py:
```diff
@@ -0,0 +1,57 @@
+from ml_model_providers.models import ModelProviderConnection, ModelProviders
+
+
+def _fill_model_version_model_provider_connection():
+    third_party_model_versions = ThirdPartyModelVersion.objects.all()
+    for model_version in third_party_model_versions:
+        if model_version.provider == ModelProviders.OPENAI:
+            with transaction.atomic():
+                model_provider_connection = ModelProviderConnection.objects.filter(
+                    organization=model_version.organization,
+                    provider=model_version.provider
+                ).first()
+                model_version.model_provider_connection = model_provider_connection
+                model_version.save()
+        elif model_version.provider == ModelProviders.AZURE_OPENAI:
+            with transaction.atomic():
+                model_provider_connection = ModelProviderConnection.objects.filter(
+                    organization=model_version.organization,
+                    provider=model_version.provider,
+                    deployment_name=model_version.provider_model_id,
+                ).first()
+                model_version.model_provider_connection = model_provider_connection
+                model_version.save()
```
```suggestion
def _fill_model_version_model_provider_connection():
    for provider in [ModelProviders.OPENAI, ModelProviders.AZURE_OPENAI]:
        this_provider_model_versions = ThirdPartyModelVersion.objects.filter(provider=provider).values('id', 'organization_id', 'provider_model_id')
        for provider_model_version in this_provider_model_versions:
            connection_ids = ModelProviderConnection.objects.filter(
                organization_id=provider_model_version['organization_id'],
                provider=provider,
                **({'deployment_name': provider_model_version['provider_model_id']} if provider == ModelProviders.AZURE_OPENAI else {}),
            ).values_list('id')[:1]
            connection_id = connection_ids[0] if connection_ids else None
            ThirdPartyModelVersion.objects.filter(id=provider_model_version['id']).update(model_provider_connection_id=connection_id)
     ```

jombooth replied:
^ please test this because I wrote it without running the code, but the general idea is not to instantiate any ModelProviderConnection or ThirdPartyModelVersion objects and select/update only the exact fields that we need to.

Commit:
jo-ify the async migration

jombooth commented on label_studio/ml_models/migrations/0011_thirdpartymodelversion_model_provider_connection.py:
```diff
@@ -0,0 +1,49 @@
+                **({'deployment_name': provider_model_version['provider_model_id']} if provider == ModelProviders.AZURE_OPENAI else {}),
+            ).values_list('id', flat=True)[:1]
+            connection_id = connection_ids[0] if connection_ids else None
+            print(f"id: {provider_model_version['id']}")
+            print(f"mpc id: {connection_id}")
```
nit: I usually favor using `logger.info`

hakan458 replied:
accidentally left them in, removed

Commit:
remove print statements
